core/src/main/scala/za/co/absa/spline/harvester/postprocessing/AttributeOrderEnrichingFilter.scala

```diff
@@ -0,0 +1,97 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spline.harvester.postprocessing
+
+import org.apache.commons.configuration.Configuration
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import za.co.absa.commons.reflect.ReflectionUtils._
+import za.co.absa.commons.reflect.extractors.SafeTypeMatchingExtractor
+import za.co.absa.spline.harvester.HarvestingContext
+import za.co.absa.spline.harvester.builder.UnionNodeBuilder.ExtraFields
+import za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin.`_: V2WriteCommand`
+import za.co.absa.spline.producer.model.v1_1.{DataOperation, ExecutionPlan, WriteOperation}
+
+import java.util.UUID
+
+class AttributeOrderEnrichingFilter(conf: Configuration) extends AbstractPostProcessingFilter {
+
+  override def processExecutionPlan(plan: ExecutionPlan, ctx: HarvestingContext): ExecutionPlan = {
+    val isByName = plan
+      .operations.write.params
+      .flatMap(_.get("isByName").map(_.asInstanceOf[Boolean]))
+      .getOrElse(false)
+
+    if (isByName)
+      addSyntheticReorderingSelect(plan, ctx)
+    else
+      plan
+  }
+
+  private def addSyntheticReorderingSelect(plan: ExecutionPlan, ctx: HarvestingContext): ExecutionPlan = {
+
+    val writeOp = plan.operations.write
+    val writeChildOutput = getWriteChildOutput(plan, writeOp)
+
+    val reorderedOutput = newOrder(ctx)
+      .zip(writeChildOutput)
+      .sortBy{ case (order, attribute) => order }
+      .map{ case (order, attribute) => attribute }
+
+    val syntheticProjection = DataOperation(
+      id = UUID.randomUUID().toString,
+      name = Some("Project"),
+      childIds = Some(writeOp.childIds),
+      output = Some(reorderedOutput),
+      params = None,
+      extra = Some(Map(ExtraFields.Synthetic -> true))
+    )
+
+    plan.copy(
+      operations = plan.operations.copy(
+        write = writeOp.copy(childIds = Seq(syntheticProjection.id)),
+        other = Some(plan.operations.other.getOrElse(Seq.empty) :+ syntheticProjection)
+      )
+    )
+  }
+
+  private def newOrder(ctx: HarvestingContext): Seq[Int] = ctx.logicalPlan match {
+    case `_: V2WriteCommand`(writeCommand) => {
+      val namedRelation = extractFieldValue[AnyRef](writeCommand, "table")
+      val finalOutput = extractFieldValue[Seq[Attribute]](namedRelation, "output")
+      val query = extractFieldValue[LogicalPlan](writeCommand, "query")
+
+      query.output.map(att => finalOutput.indexWhere(_.name == att.name))
+    }
+  }
+
+  private def getWriteChildOutput(plan: ExecutionPlan, writeOp: WriteOperation): Seq[String] = {
+    val writeChildId = writeOp.childIds.head
```

Comment:
Calling `head` implies that there is a `tail` and thus it raises a question - why we only take `head`, what about the the rest items?

I would add some explanation:
```suggestion
    assume(writeOp.childIds.length == 1) // In Spark, a Write operation is expected to have a single child.
    val writeChildId = writeOp.childIds.head
```

or even better, using `unapply` suggests to me that we know the exact list structure, no tail is assumed, so all is clear.

```suggestion
    val Seq(writeChildId) = writeOp.childIds
```